p2p/src/codec.rs

```diff
@@ -0,0 +1,244 @@
+	}
+
+	/// Length of the next item we are expecting, could be msg header, body, block header or attachment chunk
+	fn next_len(&self) -> usize {
+		match &self.state {
+			None => MsgHeader::LEN,
+			Header(Known(h)) if h.msg_type == Type::Headers => {
+				// If we are receiving a list of headers, read off the item count first
+				min(h.msg_len as usize, 2)
+			}
+			Header(Known(header)) => header.msg_len as usize,
+			Header(Unknown(len, _)) => *len as usize,
+			BlockHeaders { bytes_left, .. } => {
+				// The header length varies with the number of edge bits. Therefore we overestimate
+				// its size and only actually read the bytes we need
+				min(*bytes_left, header_size_bytes(63))
+			}
+			Attachment(left, _, _) => min(*left, 48_000),
+		}
+	}
+
+	/// Set stream timeout depending on the next expected item
+	fn set_stream_timeout(&self) -> Result<(), Error> {
+		let timeout = match &self.state {
+			None => HEADER_IO_TIMEOUT,
+			_ => BODY_IO_TIMEOUT,
+		};
+		self.stream.set_read_timeout(Some(timeout))?;
+		Ok(())
+	}
+
+	fn read_inner(&mut self) -> Result<Message, Error> {
+		self.bytes_read = 0;
+		loop {
+			let next_len = self.next_len();
+			let pre_len = self.buffer.len();
+			// Buffer could already be partially filled, calculate additional bytes we need
+			let to_read = next_len.saturating_sub(pre_len);
+			if to_read > 0 {
+				self.buffer.reserve(to_read);
+				for _ in 0..to_read {
+					self.buffer.put_u8(0);
+				}
+				self.set_stream_timeout()?;
+				if let Err(e) = self.stream.read_exact(&mut self.buffer[pre_len..]) {
+					// Undo reserved bytes on a failed read
+					self.buffer.truncate(pre_len);
+					return Err(e.into());
+				}
+				self.bytes_read += to_read;
+			}
+			match &mut self.state {
+				None => {
+					// Parse header and keep reading
+					let mut raw = self.buffer.split_to(next_len).freeze();
+					let mut reader = BufReader::new(&mut raw, self.version);
+					let header = MsgHeaderWrapper::read(&mut reader)?;
+					self.state = Header(header);
+				}
+				Header(Known(header)) => {
+					let mut raw = self.buffer.split_to(next_len).freeze();
+					if header.msg_type == Type::Headers {
+						// Special consideration for a list of headers, as we want to verify and process
+						// them as they come in instead of only after the full list has been received
+						let mut reader = BufReader::new(&mut raw, self.version);
+						let items_left = reader.read_u16()? as usize;
+						self.state = BlockHeaders {
+							bytes_left: header.msg_len as usize - 2,
+							items_left,
+							headers: Vec::with_capacity(min(HEADER_BATCH_SIZE, items_left)),
+						};
+					} else {
+						// Return full message
+						let msg = decode_message(header, &mut raw, self.version);
+						self.state = None;
+						return msg;
+					}
+				}
+				Header(Unknown(_, msg_type)) => {
+					// Discard body and return
+					let msg_type = *msg_type;
+					self.buffer.advance(next_len);
```

Comment:
This just discards those bytes by advancing over them? :+1:

Reply:
Yes, `Bytes`/`BytesMut` has an internal cursor, this just increases it by `next_len` bytes. Since this is the only place that points to this part of the buffer (guaranteed because it is a `BytesMut`), that part gets marked as dropped.

Splitting the data off like we do when we read the body ([here](https://github.com/mimblewimble/grin/pull/3433/files#diff-194510b22fabb1d5607a8bca1200a847R130)) advances the cursor similarly, but also increases a refcount and returns a `Bytes` for that part of the buffer. We don't actually care about the content of the body, so we just advance the cursor which is a bit faster.